src/dynamics/solver/velocity_constraint.rs

```diff
@@ -239,13 +239,11 @@ impl VelocityConstraint {
                             + gcross1.gdot(gcross1)
                             + gcross2.gdot(gcross2));
 
-                    let mut rhs = (vel1 - vel2).dot(&force_dir1);
-
-                    if rhs <= -params.restitution_velocity_threshold {
-                        rhs += manifold_point.restitution * rhs
-                    }
-
-                    rhs += manifold_point.dist.max(0.0) * inv_dt;
+                    let rhs = if manifold_point.is_bouncy() {
+                        (1.0 + manifold_point.restitution) * (vel1 - vel2).dot(&force_dir1)
+                    } else {
+                        (vel1 - vel2).dot(&force_dir1) + manifold_point.dist.max(0.0) * inv_dt
+                    };
```

Comment:
This assumes that all the bouncy contacts are immediately active. Instead we should include the predictive contact term `+ manifold_point.dist.max(0.0) * inv_dt` in both cases:


```rust
let effective_restitution = manifold_point.is_bouncy() as u32 as Real * manifold_point.restitution;
let rhs = (1.0 + effective_restitution) * (vel1 - vel2).dot(&force_dir1) + manifold_point.dist.max(0.0) * inv_dt;
```